clippy_lints/src/question_mark.rs

```diff
@@ -95,19 +95,25 @@ impl QuestionMark {
 
     fn check_if_let_some_or_err_and_early_return(cx: &LateContext<'_>, expr: &Expr<'_>) {
         if_chain! {
-            if let Some(higher::IfLet { let_pat, let_expr, if_then, if_else: Some(if_else) })
+            if let Some(higher::IfLet { let_pat, let_expr, if_then, if_else })
                 = higher::IfLet::hir(cx, expr);
             if let PatKind::TupleStruct(ref path1, fields, None) = let_pat.kind;
-            if (Self::option_check_and_early_return(cx, let_expr, if_else) && is_lang_ctor(cx, path1, OptionSome)) ||
-                (Self::result_check_and_early_return(cx, let_expr, if_else) && is_lang_ctor(cx, path1, ResultOk));
-
-            if let PatKind::Binding(annot, bind_id, _, _) = fields[0].kind;
+            // Only check one of the blocks
+            let nested_expr = if_else.unwrap_or(if_then);
+            if Self::check_lang_items(cx, path1, &[OptionSome, ResultOk, ResultErr]);
```

Comment:
I don't think this is correct to place it here, this will result in a different behavior compared to the original. When the let pattern is `Result` we want the return to also be `Result`, but placing it here will make `option_check_and_early_return` called for `Result`.

For example, I believe this will trigger when `check_lang_items` is here

```rust
fn test_1() -> Option<String> {
    if let Err(err) = function_returning_result() {
        return Some(err);
    }

    None
}
```

Reply:
Ahhh, I see, thank you for all the feedbacks!

Reply:
the original code kinda has this problem as well, yes, your example did trigger this lint, but I guess maybe it is because [this block here](https://github.com/rust-lang/rust-clippy/blob/1c87b2df8ae13dfb404805840c041789778310e6/clippy_lints/src/question_mark.rs#L193)? where I only check for `err` symbol inside the parentheses... 

How do I compare `Err()` and `Some()` tho? Their HirId.owner look the same but I can't directly compare them right?

Reply:
Before the changes, it's checked here

https://github.com/rust-lang/rust-clippy/blob/ea4db3a6c6fdd30abb0b746f7a68d403a07d24bf/clippy_lints/src/question_mark.rs#L101-L102

This does not trigger before ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4db6d0ba767bc645f714e97e7996db64))

Reply:
Cuz it didn't check `ExprKind::Call` return and `if without else` at all

Reply:
Whoops, that playground example I gave is wrong.

But yeah, the problem is a little different now, I think what I meant to say is that, previously we checked it with those lines and it's sufficient, but this is not the case anymore. 

For the changes, I think you can just check if the return value is an `Err()` because we matched `if let Err() ...` in the let pattern, without checking the owner.

Reply:
That helps a lot, thanks!